### DISCUSSION TITLE
Type-safe message authoring

### DISCUSSION THREAD

**Discussion by musjj** [amannn/next-intl#1267] [Q&A]
I want to have completions and type-checking while authoring my messages, so that I won't forget to include certain namespaces, etc. The docs suggests something like this:

`global.d.ts`
```typescript
import en from './messages/en.json';
 
type Messages = typeof en;
 
declare global {
  // Use type safe message keys with `next-intl`
  interface IntlMessages extends Messages {}
}
```

This is useful for type-inferring the messages you've written for a single locale, but it doesn't help ensure that your messages are uniform and correct across all locales.

I'm thinking of just using `.ts` files for my messages so that I can just directly typecheck them, like this:

`./messages/en.ts`
```typescript
import type { Messages } from "./schema.ts";

export default {
  ...
} satisfies Messages;
```

But is this a good practice? Is there a downside to this approach? I considered using JSON schemas, but I feel that would complicate the build process and code generation.

**Comment by amannn** [comment]
A downside of this approach is that unlike JSON files, messages in source code might not be compatible with other tools like [the VSCode integration for `next-intl`](https://next-intl-docs.vercel.app/docs/workflows/vscode-integration) or translation management systems like [Crowdin](https://crowdin.com/teams/engineering). Such translation management systems are also really good at helping you ensure that messages are in sync across all locales—maybe something you could consider!
